{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For Lab 7, we must compare different algorithms for effiency in compressing 1000 terabytes of data each day.  To start, we must first generate 100 megabytes of random binary data.  To do this we will use the given commands below to utilize np.random.choice.  We will also use np.packbits in order to later on write a files for this data.\n",
    "\n",
    "This method can generate ones and zeros with varying probabilities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 1 0 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 1 1 0 1 0 1 0 1 1 0 0 1 1 1 1 0 1 1 0\n",
      " 1 1 1 0 1 0 1 1 1 1 1 1 1 0 1 0 0 1 0 0 1 1 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1\n",
      " 1 0 0 1 1 1 0 1 1 1 1 1 1 0 1 1 0 0 0 0 1 1 0 0 0 0 1 1]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "myvar = np.random.choice([0, 1], size = 102, replace = True, p = [0.5, 0.5])\n",
    "print(myvar)\n",
    "myvar = np.packbits(myvar)\n",
    "open(\"myvar\", \"wb\").write(myvar)\n",
    "\n",
    "#print(myvar)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Simulating the data:**\n",
    "After becoming acquainted with how these two functions work, I can alter the probabilities of generating a one or zero to generate binary files with different percentages of zeros.\n",
    "\n",
    "I can also use the same random.choice method to produce sequences of nucleotides/amino acids with equal probabilities.  These sequences can be used to produce fasta files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "#Generating 100 megabytes of random data, containing 100%, 90%, 80%, 70%, 60%, and 50% zeros\n",
    "# 100 megabytes = 8 bits/byte * 1024 bytes/kilobyte * 1024 kilobytes/megabyte *100 = 838860800 bits\n",
    "\n",
    "#100% present zeros\n",
    "zeros_100p = np.random.choice([0, 1], size = 838860800, replace = True, p = [1.0, 0.0])\n",
    "zeros_100p = np.packbits(zeros_100p)\n",
    "\n",
    "# #90% present zeros\n",
    "# zeros_90p = np.random.choice([0, 1], size = 838860800, replace = True, p = [0.9, 0.1])\n",
    "# zeros_90p = np.packbits(zeros_90p)\n",
    "\n",
    "# #80% present zeros\n",
    "# zeros_80p = np.random.choice([0, 1], size = 838860800, replace = True, p = [0.8, 0.2])\n",
    "# zeros_80p = np.packbits(zeros_80p)\n",
    "\n",
    "# #70% present zeros\n",
    "# zeros_70p = np.random.choice([0, 1], size = 838860800, replace = True, p = [0.7, 0.3])\n",
    "# zeros_70p = np.packbits(zeros_70p)\n",
    "\n",
    "# #60% present zeros\n",
    "# zeros_60p = np.random.choice([0, 1], size = 838860800, replace = True, p = [0.6, 0.4])\n",
    "# zeros_60p = np.packbits(zeros_60p)\n",
    "\n",
    "# #50% present zeros\n",
    "# zeros_50p = np.random.choice([0, 1], size = 838860800, replace = True, p = [0.5, 0.5])\n",
    "# zeros_50p = np.packbits(zeros_50p)\n",
    "\n",
    "open(\"zeros_100p\", \"wb\").write(zeros_100p)\n",
    "# open(\"zeros_90p\", \"wb\").write(zeros_90p)\n",
    "# open(\"zeros_80p\", \"wb\").write(zeros_80p)\n",
    "# open(\"zeros_70p\", \"wb\").write(zeros_70p)\n",
    "# open(\"zeros_60p\", \"wb\").write(zeros_60p)\n",
    "# open(\"zeros_50p\", \"wb\").write(zeros_50p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "#Generating a random DNA sequence\n",
    "DNAseq = np.random.choice([\"A\", \"T\", \"C\",\"G\"], size = 100000000, replace = True, p = [0.25, 0.25, 0.25, 0.25])\n",
    "open(\"DNAseq.fa\",\"w\").write(\"\".join(DNAseq))\n",
    "\n",
    "#Generating a random RNA sequence\n",
    "AAsymbols = [\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\",\"S\", \"T\", \"V\", \"W\", \"Y\"]\n",
    "\n",
    "Proteinseq = np.random.choice(AAsymbols, size = 100000000, replace = True, p = [0.05]*20)\n",
    "open(\"Proteinseq.fa\",\"w\").write(\"\".join(Proteinseq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Results:**  \n",
    "The results below show the compression results for four algorithms on all the random files of information produced above.\n",
    "\n",
    "**Example Terminal Commands:**\n",
    "* time gzip -k zeros_100p  \n",
    "* time bzip2 -k zeros_100p  \n",
    "* time ArithmeticCompress zeros_100p zeros_100p.art  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "<h2>Zeros 100p</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>104857600</td>\n",
    "<td>101802</td>\n",
    "<td>0.727s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>113</td>\n",
    "    <td>1.060s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>5615</td>\n",
    "    <td>0.113s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>104857600</td>\n",
    "    <td>1028</td>\n",
    "    <td>14.995s</td>\n",
    "  </tr>\n",
    "</table>\n",
    "</body>\n",
    "</html>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Zeros 90p</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>104857600</td>\n",
    "<td>58737154</td>\n",
    "<td>18.833s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>61159531</td>\n",
    "    <td>10.789s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>61183812</td>\n",
    "    <td>0.738s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>104857600</td>\n",
    "    <td>49178603</td>\n",
    "    <td>28.715s</td>\n",
    "  </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Zeros 80p</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>104857600</td>\n",
    "<td>81158408</td>\n",
    "<td>13.388s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>86643159</td>\n",
    "    <td>12.032s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>86655342</td>\n",
    "    <td>0.951s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>104857600</td>\n",
    "    <td>75697967</td>\n",
    "    <td>35.472s</td>\n",
    "  </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Zeros 70p</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>104857600</td>\n",
    "<td>93618979</td>\n",
    "<td>6.085s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>99761913</td>\n",
    "    <td>13.854s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>99767682</td>\n",
    "    <td>1.128s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>104857600</td>\n",
    "    <td>92412569</td>\n",
    "    <td>39.156s</td>\n",
    "  </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Zeros 60p</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>104857600</td>\n",
    "<td>102418465</td>\n",
    "<td>4.263s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>104911427</td>\n",
    "    <td>15.712s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>104917527</td>\n",
    "    <td>1.432s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>104857600</td>\n",
    "    <td>101811899</td>\n",
    "    <td>41.027s</td>\n",
    "  </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Zeros 50p</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>104857600</td>\n",
    "<td>104874290</td>\n",
    "<td>3.554s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>105319977</td>\n",
    "    <td>16.732s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>104857600</td>\n",
    "    <td>105329147</td>\n",
    "    <td>1.507s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>104857600</td>\n",
    "    <td>104858608</td>\n",
    "    <td>7.249s</td>\n",
    "  </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>DNA Seq</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>100000000</td>\n",
    "<td>29223445</td>\n",
    "<td>15.156s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>100000000</td>\n",
    "    <td>27336271</td>\n",
    "    <td>9.491s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>100000000</td>\n",
    "    <td>27346933</td>\n",
    "    <td>0.693s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>100000000</td>\n",
    "    <td>25001027</td>\n",
    "    <td>21.471s</td>\n",
    "  </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Protein Seq</h2>\n",
    "<table text-align = \"left\">\n",
    "<tr>\n",
    "<th>Algorithm</th>\n",
    "<th>Original Size (byte)</th>\n",
    "<th>New Size (byte)</th>\n",
    "<th>Time (s) </th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>Gzip</td>\n",
    "<td>100000000</td>\n",
    "<td>60559249</td>\n",
    "<td>4.242s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Bzip2</td>\n",
    "    <td>100000000</td>\n",
    "    <td>55259521</td>\n",
    "    <td>10.282s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>Pbzip2</td>\n",
    "    <td>100000000</td>\n",
    "    <td>55269829</td>\n",
    "    <td>0.797s</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td>ArithmeticCompress</td>\n",
    "    <td>100000000</td>\n",
    "    <td>54025127</td>\n",
    "    <td>28.635s</td>\n",
    "  </tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Questions:**  \n",
    "Q. Which algorithm achieves the best level of compression on each file type?  \n",
    "A.  \n",
    "zeros_100p: Bzip2  \n",
    "zeros_90p: ArithmeticCompress  \n",
    "zeros_80p: ArithmeticCompress  \n",
    "zeros_70p: ArithmeticCompress  \n",
    "zeros_60p: ArithmeticCompress  \n",
    "zeros_50p: ArithmeticCompress  \n",
    "DNAseq: ArithmeticCompress  \n",
    "Proteinseq: ArithmeticCompress   \n",
    "\n",
    "Q. Which algorithm is the fastest?  \n",
    "A.  \n",
    "zeros_100p: Pbzip2  \n",
    "zeros_90p: Pbzip2  \n",
    "zeros_80p: Pbzip2  \n",
    "zeros_70p: Pbzip2  \n",
    "zeros_60p: Pbzip2  \n",
    "zeros_50p: Pbzip2  \n",
    "DNAseq: Pbzip2  \n",
    "Proteinseq: Pbzip2 \n",
    "\n",
    "Q. What is the difference between bzip2 and pbzip2? Do you expect one to be faster and why?  \n",
    "A.  \n",
    "Pbzip2 is a parallel implementation of bzip2 which is a block-sorting file compressor. I would expect pbzip2 to be faster since it is running multiple functions at the same time.\n",
    "\n",
    "Q. How does the level of compression change as the percentage of zeros increases? Why does this happen?  \n",
    "A.  \n",
    "The level of compression increases as the percentage of zeros increases.  This is because the more zeros there are, the less meaningful information there is in the file (versus a uniform probability of zeros and ones).\n",
    "\n",
    "Q. What is the minimum number of bits required to store a single DNA base?  \n",
    "A.  \n",
    "2 bits are required to store a single DNA base.\n",
    "\n",
    "Q. What is the minimum number of bits required to store an amino acid letter?  \n",
    "A.  \n",
    "5 bits are required to store a single amino acid.\n",
    "\n",
    "Q. In your tests, how many bits did gzip and bzip2 actually require to store your random DNA and protein sequences?  \n",
    "A.  \n",
    "DNA  \n",
    "Gzip: 29223445B * 8bits/KB = 233,787,560bits  \n",
    "Bzip2: 27336271B * 8bits/KB = 218,690,168bits  \n",
    "Protein  \n",
    "Gzip: 60559249B * 8bits/KB = 484,473,992bits  \n",
    "Bzip2: 55259521B * 8bits/KB = 442,076,168bits  \n",
    "\n",
    "Q. Are gzip and bzip2 performing well on DNA and proteins?  \n",
    "A.  \n",
    "The level of compression is roughly half, which is relatively good compared to the other files.  This also considers that the original file represented each nucleotide/acid as a byte."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Compressing real data:** My real data consists of 10 homologs of gp120 from different HIV isolates.  I will find these searching \"gp120\" in the ncbi nucleotide database.  In order to keep my seqeunces consistent, I chose to search for the whole gp120 gene and applied a search filter for sequence length (1300-1500 bps).  I downloaded the fasta files directly from the website and will make a multifasta file below:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "from Bio import SeqIO\n",
    "\n",
    "records=[]\n",
    "\n",
    "seq1 = open(\"gp120seq1.fasta\")\n",
    "seq2 = open(\"gp120seq2.fasta\")\n",
    "seq3 = open(\"gp120seq3.fasta\")\n",
    "seq4 = open(\"gp120seq4.fasta\")\n",
    "seq5 = open(\"gp120seq5.fasta\")\n",
    "seq6 = open(\"gp120seq6.fasta\")\n",
    "seq7 = open(\"gp120seq7.fasta\")\n",
    "seq8 = open(\"gp120seq8.fasta\")\n",
    "seq9 = open(\"gp120seq9.fasta\")\n",
    "seq10 = open(\"gp120seq10.fasta\")\n",
    "\n",
    "\n",
    "records.append(SeqIO.read(seq1, \"fasta\"))\n",
    "records.append(SeqIO.read(seq2, \"fasta\"))\n",
    "records.append(SeqIO.read(seq3, \"fasta\"))\n",
    "records.append(SeqIO.read(seq4, \"fasta\"))\n",
    "records.append(SeqIO.read(seq5, \"fasta\"))\n",
    "records.append(SeqIO.read(seq6, \"fasta\"))\n",
    "records.append(SeqIO.read(seq7, \"fasta\"))\n",
    "records.append(SeqIO.read(seq8, \"fasta\"))\n",
    "records.append(SeqIO.read(seq9, \"fasta\"))\n",
    "records.append(SeqIO.read(seq10, \"fasta\"))\n",
    "\n",
    "SeqIO.write(records, \"gp120seqs.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "My new multifasta file contains all 10 homologs.  Using puTTy to ssh in to bioe131.com, I can use the compression algorithms on this new data file.\n",
    "\n",
    "Q. A priori, do you expect to achieve better or worse compression here than random data? Why?  \n",
    "A. I expect to see better compression in the real data because the nucleotides located near each other are more likely to be the same.  There might be repeat seqeunces within homologs and similarities across homologs.  For these reasons, it might take less bits to encode this information.\n",
    "\n",
    "**Commands:**\n",
    "* time gzip -k gp120seqs.fasta  \n",
    "* time bzip2 -k gp120seqs.fasta  \n",
    "* time ArithmeticCompress gp120seqs.fasta gp120seq.art  \n",
    "\n",
    "**Output:**  \n",
    "be131-24@meowth:~$ time gzip -k gp120seqs  \n",
    "\n",
    "real    0m0.002s  \n",
    "user    0m0.002s  \n",
    "sys     0m0.001s  \n",
    "be131-24@meowth:~$ time bzip2 -k gp120seqs\n",
    "\n",
    "real    0m0.004s  \n",
    "user    0m0.000s  \n",
    "sys     0m0.004s  \n",
    "be131-24@meowth:~$ mv gp120seqs.bz2 gp120seqs_bzip2.bz2 \n",
    "\n",
    "be131-24@meowth:~$ time pbzip2 -k gp120seqs\n",
    "\n",
    "real    0m0.004s  \n",
    "user    0m0.004s  \n",
    "sys     0m0.000s  \n",
    "be131-24@meowth:~$ time ArithmeticCompress gp120seqs gp120seqs.art  \n",
    "\n",
    "real    0m0.015s  \n",
    "user    0m0.010s  \n",
    "sys     0m0.005s  \n",
    "\n",
    "\n",
    "**Size Checks:**  \n",
    "\n",
    "be131-24@meowth:~$ ls -l gp120seqs  \n",
    "-rw-rw-r-- 1 be131-24 be131-24 15461 Oct 10 22:49 gp120seqs  \n",
    "be131-24@meowth:~$ ls -l gp120seqs.gz  \n",
    "-rw-rw-r-- 1 be131-24 be131-24 4007 Oct 10 22:49 gp120seqs.gz  \n",
    "be131-24@meowth:~$ ls -l gp120seqs_bzip2.bz2  \n",
    "-rw-rw-r-- 1 be131-24 be131-24 3940 Oct 10 22:49 gp120seqs_bzip2.bz2  \n",
    "be131-24@meowth:~$ ls -l gp120seqs.bz2  \n",
    "-rw-rw-r-- 1 be131-24 be131-24 3940 Oct 10 22:49 gp120seqs.bz2  \n",
    "be131-24@meowth:~$ ls -l gp120seqs.art  \n",
    "-rw-rw-r-- 1 be131-24 be131-24 6020 Oct 10 22:51 gp120seqs.art  \n",
    "\n",
    "Q. How does the compression ratio of this file compare to random data?  \n",
    "A.  Compression ratio is roughly 25%, which is slightly less than the numbers for the random data (27-29%).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Estimating compression of 1000 terabyte:**\n",
    "Q. Given the benchmarking data you obtained in this lab, which algorithm do you propose to use \n",
    "for each type of data? Provide an estimate for the fraction of space you can save using your \n",
    "compression scheme. How much of a bonus do you anticipate receiving this year? \n",
    "\n",
    "A. I would propose to use the pbzip2 algorithm.  It does not have the greatest compression (ArithmeticCompress) but it has the quickest processing and a comparable compression.  \n",
    "\n",
    "Random data to compress: 1000 terabytes of 80% genomes, 10% protein sequences, and 10% binary data.\n",
    "\n",
    "*Assuming that each compression is run once at the end of the day*  \n",
    "\n",
    "Compression efficiency of DNA: 27336271/100000000 = 27.3%  (0.693s)  \n",
    "Compression efficiency of Protein seq: 55269829/100000000 = 55.3% (0.797s)   \n",
    "Compression efficiency of Binary data: 105329147/104857600 = 100.4%  (1.507s)  \n",
    "\n",
    "Estimated Compression:80% * 27.3% + 10% * 55.3% + 10% * 100.4% = 37.4% (to 3 sigfigs)\n",
    "Estimated Fraction of Saved Space: 1-37.4% = 62.6%\n",
    "Estimated Bonus: 62.6% * $500/(percent reduction)= $31,300\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
